src/agent/protocols/build.rs

```diff
@@ -3,29 +3,152 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-use std::fs;
-use ttrpc_codegen::{Codegen, Customize};
+use std::fs::File;
+use std::io::{BufRead, BufReader, Read, Write};
+use std::path::Path;
+use std::process::exit;
+use ttrpc_codegen::{Codegen, Customize, ProtobufCustomize};
+
+fn replace_text_in_file(file_name: &str, from: &str, to: &str) -> Result<(), std::io::Error> {
+    let mut src = File::open(file_name)?;
+    let mut contents = String::new();
+    src.read_to_string(&mut contents).unwrap();
+    drop(src);
+
+    let new_contents = contents.replace(from, to);
+
+    let mut dst = File::create(&file_name)?;
+    dst.write_all(new_contents.as_bytes())?;
+
+    Ok(())
+}
+
+fn use_serde(protos: &[&str], out_dir: &Path) -> Result<(), std::io::Error> {
+    protos
+        .iter()
+        .try_for_each(|f: &&str| -> Result<(), std::io::Error> {
+            let proto_filename = Path::new(f)
```

Comment:
These few lines are a bit complicated, how about 
```rust
    let out_file = Path::new(f)
        .file_name()
        .and_then(|s| s.to_str())
        .ok_or("failed to get proto file name")
        .map(|s| {
            let t = s.replace(".proto", ".rs");
            out_dir.join(t)
        })
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
```

Reply:
Thanks @liubin - I've added that code and also added you as a contributor. Note that I had to tweak it a bit to convert the `PathBuf` back into a string for the call to `replace_text_in_file()`.